### PR TITLE
Implement SYCLSharedUSMSpace

### PR DESCRIPTION
**Pull request by masterleinad**
Most of the implementation was adapted from CudaUVMSpace.

### CODE REVIEW

File: core/src/SYCL/Kokkos_SYCL_Space.cpp [hunk 1]
```diff
@@ -345,14 +504,33 @@ SharedAllocationRecord<Kokkos::Experimental::SYCLDeviceUSMSpace,
...
+  using Header = SharedAllocationHeader;
+  using RecordSYCL =
+      SharedAllocationRecord<Kokkos::Experimental::SYCLSharedUSMSpace, void>;
+
+  Header* const h =
+      alloc_ptr ? reinterpret_cast<Header*>(alloc_ptr) - 1 : nullptr;
+
+  if (!alloc_ptr || h->m_record->m_alloc_ptr != h) {
+    Kokkos::Impl::throw_runtime_exception(std::string(
+        "Kokkos::Impl::SharedAllocationRecord<"
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Space.cpp:527]: Same comment about the whitespace

File: core/src/SYCL/Kokkos_SYCL_DeepCopy.hpp [hunk 2]
```diff
@@ -131,6 +131,157 @@ struct DeepCopy<Kokkos::Experimental::SYCLDeviceUSMSpace, Kokkos::HostSpace,
...
+            dst, src, n);
+  }
+  DeepCopy(const Kokkos::Experimental::SYCL& instance, void* dst,
+           const void* src, size_t n) {
+    (void)
+        DeepCopy<Experimental::SYCLDeviceUSMSpace,
+                 Experimental::SYCLDeviceUSMSpace, Kokkos::Experimental::SYCL>(
+            instance, dst, src, n);
+  }
+};
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_DeepCopy.hpp:150]: Did you consider doing
```C++
  template <>
  struct DeepCopy<Experimental::SYCLSharedUSMSpace,
                  Experimental::SYCLSharedUSMSpace, Kokkos::Experimental::SYCL>
      : DeepCopy<Experimental::SYCLDeviceUSMSpace,
                 Experimental::SYCLDeviceUSMSpace, Kokkos::Experimental::SYCL> {
  }

Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_DeepCopy.hpp]: Ok. Delegating makes it a few lines shorter.

File: core/src/SYCL/Kokkos_SYCL_Space.cpp [hunk 3]
```diff
@@ -145,7 +147,39 @@ void* SYCLDeviceUSMSpace::impl_allocate(
...
+    const size_t arg_logical_size,
+    const Kokkos::Tools::SpaceHandle arg_handle) const {
+  const cl::sycl::queue& queue =
+      *SYCL().impl_internal_space_instance()->m_queue;
+  void* const hostPtr = cl::sycl::malloc_shared(arg_alloc_size, queue);
+
+  if (hostPtr == nullptr) {
+    throw RawMemoryAllocationFailure(
+        arg_alloc_size, 1, RawMemoryAllocationFailure::FailureMode::Unknown,
+        RawMemoryAllocationFailure::AllocationMechanism::SYCLMallocShared);
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Space.cpp:181]: Yes.

File: core/src/Kokkos_SYCL_Space.hpp [hunk 4]
```diff
@@ -173,6 +260,56 @@ class SharedAllocationRecord<Kokkos::Experimental::SYCLDeviceUSMSpace, void>
                             bool detail = false);
 };
 
+template <>
+class SharedAllocationRecord<Kokkos::Experimental::SYCLSharedUSMSpace, void>
+    : public SharedAllocationRecord<void, void> {
+ private:
+  using RecordBase = SharedAllocationRecord<void, void>;
+
+  SharedAllocationRecord(const SharedAllocationRecord&) = delete;
```
Comment (nliber) [inline core/src/Kokkos_SYCL_Space.hpp:269]: We should delete the move operations too (Rule of 5).

File: core/src/SYCL/Kokkos_SYCL_Space.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_Space.cpp:181]: There is a more general `sycl::malloc` call which takes a `enum class sycl::usm::kind` to indicate the type of memory, if that would help in eliminating some of the code duplication.

File: core/src/SYCL/Kokkos_SYCL_Space.cpp [hunk 5]
```diff
@@ -227,6 +290,33 @@ SharedAllocationRecord<Kokkos::Experimental::SYCLDeviceUSMSpace, void>::
...
+      m_space(arg_space) {
+  // Fill in the Header information
+  RecordBase::m_alloc_ptr->m_record = this;
+
+  strncpy(RecordBase::m_alloc_ptr->m_label, arg_label.c_str(),
+          SharedAllocationHeader::maximum_label_length);
+
+  // Set last element zero, in case c_str is too long
+  RecordBase::m_alloc_ptr
+      ->m_label[SharedAllocationHeader::maximum_label_length - 1] = (char)0;
```
Comment (nliber) [inline core/src/SYCL/Kokkos_SYCL_Space.cpp:317]: I usually write `(char)0` as `'\0'`, mainly because I really don't like to cast.

**Review by nliber — APPROVED** [review]

File: core/src/SYCL/Kokkos_SYCL_Space.cpp (hunk 5 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Space.cpp]: Done.

File: core/src/Kokkos_SYCL_Space.hpp (hunk 4 above)
Comment (masterleinad) [reply core/src/Kokkos_SYCL_Space.hpp:253]: Done.

File: core/src/SYCL/Kokkos_SYCL_Space.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL_Space.cpp]: Yes, I tried a bit harder and could remove some code duplications.

**Comment by masterleinad** [comment]
I addressed @nliber's comments.

**Review by nliber — APPROVED** [review]

**Review by crtrott — CHANGES_REQUESTED** [review]
I think in principle we want allocate and deallocate to be member functions not static functions. While they could be static functions now, it could lead to usage we need to fix later. One of the scenarios behind this is that memory spaces could be backed by memory pools which could be per instance, or we end up supporting multiple devices in a single process and thus making the calls instance specific.

**Comment by masterleinad** [comment]
@crtrott Here you go!

**Review by dalg24 — APPROVED** [review]

**Comment by masterleinad** [comment]
`Travis` seems to be stuck but we are only touching `SYCL` anyway.

**Review by crtrott — APPROVED** [review]

**Pull request merged**